webapp-django/crashstats/symbols/management/commands/reencode-bad-sym-files.py

```diff
@@ -0,0 +1,132 @@
+        ),
+        make_option(
+            '--max-uploads',
+            dest='max_uploads',
+            default='0',
+            help='Max. number of SymbolsUploads to work on (default all)'
+        ),
+    )
+
+    def _parse_date(self, date):
+        year, month, day = [int(x) for x in date.split('-')]
+        return timezone.now().replace(
+            year=year,
+            month=month,
+            day=day,
+            hour=0,
+            minute=0,
+            second=0
+        )
+
+    def handle(self, *args, **options):
+        first_date = self._parse_date(options['first_date'])
+        end_date = self._parse_date(options['end_date'])
+        max_uploads = int(options['max_uploads'])
+        uploads = SymbolsUpload.objects.filter(
+            created__gte=first_date,
+            created__lt=end_date
+        )
+        buckets = {}
+        conn = boto.connect_s3(
+            settings.AWS_ACCESS_KEY,
+            settings.AWS_SECRET_ACCESS_KEY,
+            calling_format=boto.s3.connection.OrdinaryCallingFormat()
+        )
+        if max_uploads > 0:
+            uploads = uploads.order_by('?')[:max_uploads]
+        for upload in uploads:
+            print repr(upload)
+            for line in upload.content.splitlines():
+                if not line.endswith('.sym'):
+                    continue
+                bucket_name = line.split(',')[0]
+                if bucket_name.startswith('+') or bucket_name.startswith('='):
+                    bucket_name = bucket_name[1:]
+                if bucket_name not in buckets:
+                    buckets[bucket_name] = conn.lookup(bucket_name)
+
+                if not buckets[bucket_name]:
+                    print "No bucket called", repr(bucket_name)
+                    continue
+
+                key = buckets[bucket_name].get_key(line.split(',')[1])
+                if key is not None:
+                    assert key.content_type == 'text/plain', key.content_type
+                    assert key.content_encoding == 'gzip', key.content_encoding
+                    print filesizeformat(key.size).ljust(10),
+                    url = key.generate_url(expires_in=0, query_auth=False)
+                    print url
+                    if self._correctly_encoded(url):
+                        print "CORRECTLY ENCODED"
+                    else:
+                        print "NOT CORRECT",
+                        if not options['dry_run']:
+                            downloaded = key.get_contents_as_string()
+                            real_content = downloaded.decode('zlib')
+                            out = StringIO()
+                            with gzip.GzipFile(fileobj=out, mode='w') as f:
+                                f.write(real_content)
+                            value = out.getvalue()
+                            key.set_contents_from_string(value, {
+                                'Content-Encoding': 'gzip'
+                            })
```

Comment:
I don't know they that deflate thing works but it feels like a hack on top of a hack. We should encode the files properly.